btx/processing/pipca.py

```diff
@@ -0,0 +1,1011 @@
+            self.num_components,
+            self.batch_size,
+            self.num_features,
+        ) = self.set_params(num_images, num_components, batch_size, bin_factor)
+
+        # compute number of counts in and start indices over ranks
+        self.split_indices, self.split_counts = distribute_indices(
+            self.num_features, self.size
+        )
+
+        # attribute for storing interval data
+        self.task_durations = dict({})
+
+        self.num_incorporated_images = 0
+        self.outliers, self.pc_data = [], []
+
+    def get_params(self):
+        """
+        Method to retrieve iPCA params.
+
+        Returns
+        -------
+        num_incorporated_images : int
+            number of images used to build model
+        num_components : int
+            number of components maintained in model
+        batch_size : int
+            batch size used in model updates
+        num_features : int
+            dimensionality of incorporated images
+        """
+        return (
+            self.num_incorporated_images,
+            self.num_components,
+            self.batch_size,
+            self.num_features,
+        )
+
+    def set_params(self, num_images, num_components, batch_size, bin_factor):
+        """
+        Method to initialize iPCA parameters.
+
+        Parameters
+        ----------
+        num_images : int
+            Desired number of images to incorporate into model.
+        num_components : int
+            Desired number of components for model to maintain.
+        batch_size : int
+            Desired size of image block to be incorporated into model at each update.
+        bin_factor : int
+            Factor to bin data by.
+
+        Returns
+        -------
+        n : int
+            Number of images to incorporate into model.
+        q : int
+            Number of components for model to maintain.
+        m : int
+            Size of image block to be incorporated into model at each update.
+        d : int
+            Number of features (dimension) in each image.
+        """
+        max_events = self.psi.max_events
+        benchmarking = self.benchmarking
+        downsample = self.downsample
+
+        # set n, q, and m
+        n = num_images
+        q = num_components
+        m = batch_size
+
+        if benchmarking:
+            min_n = max(int(4 * q), 40)
+            n = min(min_n, max_events)
+            m = q
+
+            print(f"In benchmarking mode, setting q = {q}, n = {n}, m = {m}.")
+        else:
+            n = min(n, max_events) if n != -1 else max_events
+            q = min(q, n)
+            m = min(m, n)
+
+        # set d
+        det_shape = self.psi.det.shape()
+        d = np.prod(det_shape).astype(int)
+
+        if downsample:
+            if det_shape[-1] % bin_factor or det_shape[-2] % bin_factor:
+                print("Invalid bin factor, toggled off downsampling.")
+                self.downsample = False
```

Comment:
good catch but maybe we could round it to the closest valid bin factor?

Reply:
I think the argument at the time was that the user is assumed to be providing valid inputs related to the binning of data. Defaulting to closest valid bin factor is another approach but will probably be nontrivial to implement. Maybe we should add it as a feature after the PR?

Reply:
Sounds good!